Adds server api to retrieve android build variants

chunhtai opened the pull request:
as title.

dds will also need to be updated https://github.com/dart-lang/sdk/pull/53354

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or there is a reason for not adding tests.


![build.yaml badge]

If you need help, consider asking for help on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/devtools/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[build.yaml badge]: https://github.com/flutter/devtools/actions/workflows/build.yaml/badge.svg

chunhtai commented:
blocked on https://github.com/flutter/flutter/pull/131009

chunhtai commented on packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:
```diff
@@ -0,0 +1,40 @@
+import '../server/server_api.dart';
+
+class DeeplinkManager {
+  static final _buildVariantJsonRegex = RegExp(r'(\[.*\])');
```
Not sure if there is a better way to run and parse flutter command. Please advice

chunhtai commented on packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:
```diff
@@ -0,0 +1,40 @@
+    required ServerApi api,
+  }) async {
+    final ProcessResult result = await runProcess(
+      'flutter',
```
I think this is dangerous, is there a safer way to run flutter binary?

kenzieschmoll replied:
What makes this dangerous? @christopherfujino do you have an opinion here?

chunhtai replied:
This assume the flutter is in the path environment variable.

chunhtai commented on packages/devtools_app/lib/src/shared/config_specific/server/_server_web.dart:
```diff
@@ -402,6 +402,25 @@ Future<ExtensionEnabledState> extensionEnabledState({
   return ExtensionEnabledState.error;
 }
 
+Future<List<String>> requestAndroidBuildVariants(String path) async {
```
I can't find a way to reasonably test this function.

If we really want to test it, we need a way to mock  `isDevToolsServerAvailable`.
also will the ci run the test with web platform?

kenzieschmoll replied:
I don't think we have a way to test this currently. At first, I thought we might be able to test this with an integration test, but I'm not 100% confident that we start the server for those either 🤔

chunhtai commented:
Also, how can I merge this pr? I assume we will need to publish a new version and then the dds can bump the version and migrate, but I am not too familiar with dev tool release process.

Also the flutter command is added in https://github.com/flutter/flutter/pull/131009 which is not yet roll into devtool

kenzieschmoll commented on packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:
```diff
@@ -0,0 +1,40 @@
+// Copyright 2021 The Chromium Authors. All rights reserved.
```
2023

kenzieschmoll commented on packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:
```diff
@@ -0,0 +1,40 @@
+import '../server/server_api.dart';
+
+class DeeplinkManager {
+  static final _buildVariantJsonRegex = RegExp(r'(\[.*\])');
```
Can you give an example in dart doc of what the output from the process run looks like and what the result of this regex will be?

kenzieschmoll replied:
You could also add tests that test this regex matching based on example process output.

kenzieschmoll commented on packages/devtools_shared/lib/src/server/server_api.dart:
```diff
@@ -32,7 +33,8 @@ class ServerApi {
   static FutureOr<shelf.Response> handle(
     shelf.Request request,
-    ExtensionsManager extensionsManager, [
+    ExtensionsManager extensionsManager,
+    DeeplinkManager deeplinkManager, [
     ServerApi? api,
   ]) {
```
This will be a breaking change: https://github.com/dart-lang/sdk/blob/main/pkg/dds/lib/src/devtools/handler.dart#L134.

If we need to break this anyway, perhaps we should consider switching this signature to use named parameters so that we don't have to deal with more breaking changes in the future.

```
  static FutureOr<shelf.Response> handle(
    shelf.Request request, {
    required ExtensionsManager extensionsManager,
    required DeeplinkManager deeplinkManager,
    ServerApi? api,
  }) {
```

And then we'll need to do a major version bump for this package to ^4.0.0

kenzieschmoll commented on packages/devtools_shared/lib/src/server/server_api.dart:
```diff
@@ -355,3 +375,23 @@ abstract class _ExtensionsApiHandler {
+
+    final rootPath = queryParams[DeeplinkApi.deeplinkRootPathPropertyName]!;
+
+    return deeplinkManager.getBuildVariants(rootPath: rootPath, api: api);
```
instead of having `getBuildVariants` return `api.getCompleted`, let's have `getBuildVariants` return the output json, either encoded as a string or as a raw map, and then return `api.getCompleted` here. This will remove the need for `devtools_deeplink.dart` to have to depend on the server_api and will be consistent with the other api endpoints.

kenzieschmoll commented on packages/devtools_shared/test/server/server_api_test.dart:
```diff
@@ -0,0 +1,70 @@
+// Copyright 2022 The Chromium Authors. All rights reserved.
```
2023

kenzieschmoll commented on packages/devtools_shared/test/deeplink/deeplink_manager_test.dart:
```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
```
this looks like the wrong copy right. Please use the Chromium copyright

chunhtai commented:
I addressed all comment. Still unsure about https://github.com/flutter/devtools/pull/6089#discussion_r1306200951, @kenzieschmoll is there a way to locate the flutter binary?

kenzieschmoll commented on packages/devtools_shared/CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+# 4.0.0
+
+- Adds a `DeeplinkApi.androidBuildVariants` endpoint to ServerAPi.
```
nit: r/ServerAPi/ServerApi

kenzieschmoll commented on packages/devtools_shared/CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+
+- Adds a `DeeplinkApi.androidBuildVariants` endpoint to ServerAPi.
+- **BREAKING CHANGE**:
+  - ServerApi.handle extensionsManager, api parameter are converted to named parameters
```
nit: add tics and reword a bit
"`ServerApi.handle` parameters `extensionsManager` and `api` were converted to be named parameters"

kenzieschmoll commented on packages/devtools_shared/CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+- Adds a `DeeplinkApi.androidBuildVariants` endpoint to ServerAPi.
+- **BREAKING CHANGE**:
+  - ServerApi.handle extensionsManager, api parameter are converted to named parameters
+  - Adds a new required name parameter deeplinkManager to `ServerApi.handle`.
```
nit: r/name/named
and add backtics around `deepLinkManager`

kenzieschmoll commented on packages/devtools_shared/CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+# 4.0.0
```
there was a recent commit to master for version 3.0.2. It hasn't been released yet, so can you merge and move the entry from 3.0.2 to 4.0.0? and then you can remove the 3.0.2 section

kenzieschmoll replied:
This comment still stands

kenzieschmoll commented on packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:
```diff
@@ -0,0 +1,58 @@
+      };
+    }
+    return <String, String?>{
+      kErrorField: null,
```
we shouldn't need to explicitly pass null here.

kenzieschmoll commented on packages/devtools_shared/lib/src/server/server_api.dart:
```diff
@@ -355,3 +375,28 @@ abstract class _ExtensionsApiHandler {
+      api.serverError(error);
+    }
+    return api
+        .getCompleted(result[DeeplinkManager.kOutputJsonField]! as String);
```
nit: for nicer formatting, add a trailing comma after `as String`

kenzieschmoll commented:
> Also, how can I merge this pr? I assume we will need to publish a new version and then the dds can bump the version and migrate, but I am not too familiar with dev tool release process.

One this PR merges, we will need to do a release into the Dart SDK. This should be a lot simpler now, since https://dart-review.googlesource.com/c/recipes/+/323620 and have landed. 

With those changes, the upload of the DevTools build to CIPD should happen on merge of this PR. And then you'll just run `dart ./tool/update_sdk_deps.dart -c <commit-hash-for-this-pr>` to create a CL updating the devtools hash in DEPS. In that same CL, you'll want to bump the DevTools version in the `dds/pubspec.yaml` file.

sortie commented:
Yep the new devtools cipd package infra is running perfectly, I saw it make a package 2 min 46 sec after merge :)
